testing/xfpga/test_umsg_c.cpp

```diff
@@ -266,13 +303,76 @@ TEST_P(umsg_c_p, get_umsg_ptr_ioctl_err_02) {
   EXPECT_EQ(FPGA_INVALID_PARAM, xfpga_fpgaGetUmsgPtr(handle_, &value));
 }
 
+/**
+ * @test       umsg_c_p
+ * @brief      get_umsg_ptr_ioctl_err_03
+ * @details    When the parameters are valid and the drivers are loaded,
+ *             but the ioctl fails on FPGA_PORT_DMA_MAP. fpgaGetUmsgPtr returns 
+ *             FPGA_INVALID_PARAM/FPGA_EXCEPTION
+ *
+ */
 TEST_P(umsg_c_p, get_umsg_ptr_ioctl_err_03) {
   uint64_t *value = 0;
-  system_->register_ioctl_handler(FPGA_PORT_UMSG_DISABLE, dummy_ioctl<0,EINVAL>);
+
+  // register an ioctl handler that will return -1 and set errno to EINVAL
+  system_->register_ioctl_handler(FPGA_PORT_DMA_MAP, dummy_ioctl<-1,EINVAL>);
+  EXPECT_EQ(FPGA_INVALID_PARAM, xfpga_fpgaGetUmsgPtr(handle_, &value));
+
+  // register an ioctl handler that will return -1 and set errno to EFAULT
+  system_->register_ioctl_handler(FPGA_PORT_DMA_MAP, dummy_ioctl<-1,EFAULT>);
   EXPECT_EQ(FPGA_INVALID_PARAM, xfpga_fpgaGetUmsgPtr(handle_, &value));
 }
 
+/**
+ * @test       umsg_c_p
+ * @brief      invalid_free_umsg_buffer
+ * @details    When the drivers are loaded and handle umsg_virt is mapped,
+ *             but ioctl fails on FPGA_PORT_UMSG_DISABLE, FPGA_ERR outputs
+ *             "Failed to disable UMSG" and returns FPGA_OK
+ *             When ioctl fails on FPGA_PORT_UMSG_SET_BASE_ADDR, FPGA_ERR outputs
+ *             "led to zero UMSG address" and returns FPGA_OK
+ *
+ */
+TEST_P(umsg_c_p, invalid_free_umsg_buffer) {
+  uint64_t* umsg_ptr = NULL;
+  uint64_t *value = 0;
+  system_->register_ioctl_handler(FPGA_PORT_UMSG_SET_BASE_ADDR, umsg_set_base_addr);
+  system_->register_ioctl_handler(FPGA_PORT_UMSG_ENABLE, dummy_ioctl<0,EINVAL>);
```

Comment:
Did you mean to set errno to EINVAL here if returning 0?

Reply:
yeah,